k9mail/src/main/java/com/fsck/k9/activity/UnreadWidgetConfiguration.java

```diff
@@ -41,58 +64,164 @@ public void onCreate(Bundle icicle) {
+        unreadFolderEnabled.setEnabled(false);
+        unreadFolder.setEnabled(false);
+        selectedFolderName = null;
+    }
+
+    private void handleRegularAccount() {
+        Account selectedAccount = Preferences.getPreferences(this).getAccount(selectedAccountUuid);
+        unreadAccount.setSummary(selectedAccount.getName());
+        unreadFolderEnabled.setEnabled(true);
+        unreadFolder.setEnabled(true);
+    }
+
+    private void handleChooseFolder(String folderName) {
+        selectedFolderName = folderName;
+        unreadFolder.setSummary(selectedFolderName);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        getMenuInflater().inflate(R.menu.unread_widget_option, menu);
         return true;
     }
 
     @Override
-    protected void onAccountSelected(BaseAccount account) {
-        // Save widget configuration
-        String accountUuid = account.getUuid();
-        saveAccountUuid(this, mAppWidgetId, accountUuid);
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
+            case R.id.done:
+                if (validateWidget()) {
+                    updateWidgetAndExit();
+                }
+                return true;
+            default:
+                return super.onOptionsItemSelected(item);
+        }
+    }
+
+    private boolean validateWidget() {
+        if (selectedAccountUuid == null) {
+            Toast.makeText(this, R.string.unread_widget_account_not_selected, Toast.LENGTH_LONG).show();
+            return false;
+        } else if (unreadFolderEnabled.isChecked() && selectedFolderName == null) {
+            Toast.makeText(this, R.string.unread_widget_folder_not_selected, Toast.LENGTH_LONG).show();
+            return false;
+        }
+        return true;
+    }
+
+    private void updateWidgetAndExit() {
+        UnreadWidgetProperties properties = new UnreadWidgetProperties(appWidgetId, selectedAccountUuid,selectedFolderName);
+        saveWidgetProperties(this, properties);
 
         // Update widget
         Context context = getApplicationContext();
         AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
-        UnreadWidgetProvider.updateWidget(context, appWidgetManager, mAppWidgetId, accountUuid);
+        UnreadWidgetProvider.updateWidget(context, appWidgetManager, properties);
 
         // Let the caller know that the configuration was successful
         Intent resultValue = new Intent();
-        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId);
+        resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
         setResult(RESULT_OK, resultValue);
         finish();
     }
 
-    private static void saveAccountUuid(Context context, int appWidgetId, String accountUuid) {
+    private static void saveWidgetProperties(Context context, UnreadWidgetProperties properties) {
```

Comment:
Changing the format of the persisted widget configuration without updating existing data will lead to widgets created with previous app versions to stop working.
Right now we have no convenient place where such migration code could live. I propose avoiding migration by changing the format in a way that doesn't invalidate old configurations:

* `unread_widget.<widgetId>` stores the account UUID like before
* `unread_widget.<widgetId>.folderName` stores the folder name